### DISCUSSION TITLE
improve aggregation performance

### DISCUSSION THREAD

**Discussion by githorse** [TanStack/table#4860] [Ideas]
I've performed an informal grouping-and-aggregation benchmark across several popular grids (MUI, [Material React Table](https://www.material-react-table.com/), [raw Tanstack](https://tanstack.com/table/v8), [Adazzle](https://github.com/adazzle/react-data-grid), and [AG Grid](https://www.ag-grid.com/)) in this [CodeSandbox example](https://codesandbox.io/p/sandbox/8t1722?file=%2Fsrc%2FMUI.tsx%3A1%2C1). 

**Rough Loading Time (ms)[^1]**

|Rows|MUI    |MRT      |Tanstack  |AG     | Adazzle  |Best Possible|
|--------|---------|-----------|------------|----------|-------------|------------------|
|1k      | 250   |500        |250        |50        |50            |50                |
|10k    | 1000 | 1000     |500        |250      |50            |50               |
|50k    | 13000|1600     |1200      |800      |150           |150            |  
|100k  |💣[^2]   |2500      |2500      |1200    |300        |300            |

Tanstack's performance is not bad, but it's roughly twice as slow as AG Grid and 10x slower than the fastest solutions. Ok, maybe the performance of the customized solution isn't really comparable to that of a general grid solution like Tanstack, but something close to whatever AG Grid is doing ought to be possible. (One simple solution might be to use [Alasql](https://codesandbox.io/p/sandbox/8t1722?file=%2Fsrc%2Fgrids%2FAlasql.tsx)).

Questions:
1) Is there anything I can do as a user of Tanstack to accelerate or optimize aggregations?
2) Are there any optimizations Tanstack can do to bring the performance of aggregations closer to AG Grid's (or, ideally, closer to my "best possible" algorithm)?

[^1]: Most grids don't have a good way to instrument the exact loading time, so these are basically eyeballed and not very accurate. Obviously they'll be different on different machines. Nevertheless the overall trends are clear enough.
[^2]:Somewhere between 50k and 100k records the MUI grid blows up and kills the whole browser tab.

**Comment by cherniavskii** [comment]
FYI regarding MUI X Data Grid – the issue was reported in https://github.com/mui/mui-x/issues/8999 and fixed by https://github.com/mui/mui-x/pull/8990
